Experiment with Elm for implementing workout form

gfontenot opened the pull request:
Trying to wrap my head around how this (fairly complex) form would work in
Yesod made me start to wonder if something like Elm wouldn't be better suited
for the job. This is a small experiment in implementing the most basic
component I'll need for inputting workout info.

I think this experiment was a huge success, and I think that I'll probably
move forward with this model. Parsing this info at the client level and then
simply sending the results to Yesod will also let me focus on an API for the
backend, which will be necessary once I build out the mobile client anyway.

This component has 3 fields:

1. A reps field. This field can take reps for compound movements by entering
   the reps separated by "+" (and any amount of whitespace).
2. The weight for the movements. This field expects an Integer
3. A dropdown for selecting the unit of measurement. This is limited to Kilos
   and Pounds, and so crashes in the event that it gets a unit of measurement
   it doesn't expect.

The result is parsed and formatted for display to the right of the fields for
the sake of demonstration.

Some improvements I could make:

- Limit input for reps to numbers, whitespace, and "+"
- Limit input for weight to numbers

Next steps:

- Build a component that can dynamically add/remove these components and
  combine the result into something more useful.
- Add data validation to ensure that each list of reps is the same length as
  the rest and also matches the list of movements in the parent.

gfontenot committed:
Experiment with Elm for implementing workout form

Trying to wrap my head around how this (fairly complex) form would work in
Yesod made me start to wonder if something like Elm wouldn't be better suited
for the job. This is a small experiment in implementing the most basic
component I'll need for inputting workout info.

I think this experiment was a huge success, and I think that I'll probably
move forward with this model. Parsing this info at the client level and then
simply sending the results to Yesod will also let me focus on an API for the
backend, which will be necessary once I build out the mobile client anyway.

This component has 3 fields:

1. A reps field. This field can take reps for compound movements by entering
   the reps separated by "+" (and any amount of whitespace).
2. The weight for the movements. This field expects an Integer
3. A dropdown for selecting the unit of measurement. This is limited to Kilos
   and Pounds, and so crashes in the event that it gets a unit of measurement
   it doesn't expect.

The result is parsed and formatted for display to the right of the fields for
the sake of demonstration.

Some improvements I could make:

- Limit input for reps to numbers, whitespace, and "+"
- Limit input for weight to numbers

Next steps:

- Build a component that can dynamically add/remove these components and
  combine the result into something more useful.
- Add data validation to ensure that each list of reps is the same length as
  the rest and also matches the list of movements in the parent.

JoelQ commented on elm-test/Main.elm:
```diff
@@ -0,0 +1,90 @@
+
+type WeightUnit
+    = Pounds
+    | Kilos
```
Nice use of types 👍

JoelQ commented on elm-test/Main.elm:
```diff
@@ -0,0 +1,90 @@
+    { reps : List Int
+    , weight : Int
+    , unitOfMeasurement : WeightUnit
+    }
```
From a data-modeling perspective, thoughts on combining the magnitude with the unit? e.g.

```elm
type Weight
  = Points Int
  | Kils Int

type alias WorkoutSet =
  { reps : List Int
  , weight : Weight
  }
```

gfontenot replied:
I modeled it this way to mimic the way it's modeled on the server, although I actually have a transformation layer to give me this exact interface. So yeah, I think it probably makes more sense to change this to your suggestion.

JoelQ commented on elm-test/Main.elm:
```diff
@@ -0,0 +1,90 @@
+parseUnit str = case str of
+    "Pounds" -> Pounds
+    "Kilos" -> Kilos
+    _ -> Debug.crash "Not a valid unit of measurement"
```
Thoughts on returning a `Result` instead of crashing? Most parsing functions in Elm do that. e.g.

```elm
parseUnit : String -> Result String WeightUnit
parseUnit str =
  case str of
    "Pounds" -> Ok Pounds
    "Kilos" -> Ok Kilos
    _ -> Err <| str ++ " is not a valid unit of measurement"
```

gfontenot replied:
Yep, I was being lazy. I should probably return a `Result` for the `parseInt` as well (or, like, just use `String.toInt`)

JoelQ commented on elm-test/Main.elm:
```diff
@@ -0,0 +1,90 @@
+            Pounds -> "lbs"
+            Kilos -> "kgs"
+    in
+       formattedReps ++ " @ " ++ formattedWeight ++ formattedUnits
```
UI question: do you need a `" "` between the weight and units?

gfontenot replied:
I never know! Not adding a space was intentional but I'm unclear on the rules around this stuff. 1/2 the time I do like "90kgs" and the other 1/2 I do "90 kgs". [Grammar Girl seems to agree with you though](http://www.quickanddirtytips.com/education/grammar/units-of-measure) so I'll do that.

JoelQ commented on elm-test/Main.elm:
```diff
@@ -0,0 +1,90 @@
+        formattedWeight = toString set.weight
+        formattedUnits = case set.unitOfMeasurement of
+            Pounds -> "lbs"
+            Kilos -> "kgs"
```
Thoughts on extracting this case statement to a function? I think a string representation of weight unit is something you might want to use in multiple places?

gfontenot replied:
Yep, that seems like a reasonable change.

JoelQ commented on elm-test/Main.elm:
```diff
@@ -0,0 +1,90 @@
+        , select [ onInput (UnitUpdated << parseUnit) ]
+            [ viewUnit set.unitOfMeasurement Pounds
+            , viewUnit set.unitOfMeasurement Kilos
+            ]
```
There's only two elements here so maybe it's not needed but thoughts on using `List.map` here?

```elm
List.map (viewUnit set.unitOfMeasurement) units
```

gfontenot replied:
where `let units = [Pounds, Kilos]`? That makes sense. I think for the sake of explicitness (and because it's just 2 elements) I'll probably leave this as is, but that's good to keep in mind for the future.

JoelQ commented on elm-test/Main.elm:
```diff
@@ -0,0 +1,90 @@
+    div []
+        [ input [type_ "text", placeholder "reps", onInput RepsUpdated ] []
+        , input [type_ "text", placeholder "weight", onInput WeightUpdated] []
+        , select [ onInput (UnitUpdated << parseUnit) ]
```
I think most browsers don't emit an "input" event when an option is selected. Instead, you probably want to listen to the "change" event. Unfortunately, `onChange` is not implemented by `Html.Events`.

It's a one-liner to write yourself though. I wrote an article about how to do so https://robots.thoughtbot.com/building-custom-dom-event-handlers-in-elm. Scroll all the way to the bottom if you just want to see the final implementation.

gfontenot replied:
Ah, shit. This does work in Safari and Chrome but doesn't work in Firefox. I had read that this was a new thing in 0.18, but I guess it's not 100% yet? I was trying to avoid needing to do the JSON decoding stuff. Oh well. Thanks for the link, I'll make this update.

JoelQ reviewed (commented):
Nice work! 👍

gfontenot committed:
Improve Weight definition in elm

Instead of storing the unit and weight as separate fields, we can combine them
into a single Weight type, which will improve the way we use them together.

gfontenot committed:
Use custom onChange for messages from select

`onInput` works in Safari and Chrome, but not in Firefox. Because of this,
it's better to use an `onChange` function. However, this function isn't
actually supplied by Html.Events, so we need to write our own.

This function should work for anything we need to throw at it.

https://robots.thoughtbot.com/building-custom-dom-event-handlers-in-elm
